mesh/mesh.go

```diff
@@ -408,40 +442,69 @@ func (msh *Mesh) pushLayersToState(ctx context.Context, from, to types.LayerID)
 			)
 			continue
 		}
-		if err := msh.pushLayer(ctx, layerID); err != nil {
+		if err := msh.pushLayer(ctx, layerID, latestVerified); err != nil {
 			msh.missingLayer.Store(layerID)
 			return err
 		}
 		if layerID == missing {
 			msh.missingLayer.Store(types.LayerID{})
 		}
 	}
+
 	return nil
 }
 
-func (msh *Mesh) pushLayer(ctx context.Context, layerID types.LayerID) error {
+func (msh *Mesh) getBlockToApply(ctx context.Context, layerID, latestVerified types.LayerID) (*types.Block, []*types.Block, error) {
 	layerBlocks, err := msh.LayerBlocks(layerID)
 	if err != nil {
-		return fmt.Errorf("failed to get layer %s: %w", layerID, err)
+		msh.WithContext(ctx).With().Error("failed to get layer blocks", layerID, log.Err(err))
+		return nil, nil, fmt.Errorf("failed to get layer blocks %s: %w", layerID, err)
 	}
 
-	validBlocks, invalidBlocks := msh.BlocksByValidity(layerBlocks)
 	var (
-		applied    *types.Block
-		notApplied = invalidBlocks
-		blocks     = types.SortBlocks(validBlocks)
+		toApply *types.Block
+		others  []*types.Block
 	)
 
-	if len(blocks) > 0 {
+	if layerID.After(latestVerified) {
```

Comment:
i think this routine will be simplified as well if we save hare output as a contextually valid block. in such case there will be no need for special case for hare